src/msg_recorder/recorder.cc

```diff
@@ -10,25 +10,122 @@
 
 namespace cris::core {
 
-MessageRecorder::MessageRecorder(const std::filesystem::path& record_dir_prefix, std::shared_ptr<JobRunner> runner)
+MessageRecorder::MessageRecorder(const RecorderConfig& recorder_config, std::shared_ptr<JobRunner> runner)
     : Base(std::move(runner))
-    , record_dir_(record_dir_prefix / RecordDirNameGenerator()) {
+    , record_dir_(recorder_config.record_dir_ / RecordDirNameGenerator())
+    , record_strand_(MakeStrand()) {
     std::filesystem::create_directories(record_dir_);
+    SetSnapshotInterval(recorder_config);
+}
+
+void MessageRecorder::SetSnapshotInterval(const RecorderConfig& recorder_config) {
+    if (recorder_config.snapshot_intervals_.empty()) {
+        return;
+    }
+    if (recorder_config.snapshot_intervals_.size() > 1) {
+        // TODO (YuzhouGuo, https://github.com/cyfitech/cris-core/issues/97) support multi-interval
+        LOG(WARNING) << __func__
+                     << "More than one single interval received, multi-interval snapshot feature has not been "
+                        "supported, only using the last interval specified";
+    }
+    snapshot_interval_    = recorder_config.snapshot_intervals_.back().interval_sec_;
+    snapshot_subdir_name_ = recorder_config.snapshot_intervals_.back().name_;
+    snapshot_thread_      = std::thread(std::bind(&MessageRecorder::SnapshotWorkerStart, this));
+}
+
+void MessageRecorder::SnapshotWorkerStart() {
+    while (!snapshot_shutdown_flag_.load()) {
+        auto before = std::chrono::system_clock::now();
+        MakeSnapshot();
+        std::unique_lock<std::mutex> lck(snapshot_mtx_);
+        snapshot_cv_.wait_until(lck, before + snapshot_interval_, [this] { return snapshot_shutdown_flag_.load(); });
+    }
+}
+
+void MessageRecorder::MakeSnapshot() {
+    std::unique_lock<std::mutex> lock(snapshot_mtx_);
+    snapshot_pause_flag_.store(false);
+    AddJobToRunner(
+        [this]() {
+            GenerateSnapshot(keep_max_);
+            snapshot_pause_flag_.store(true);
+            snapshot_cv_.notify_all();
```

Comment:
Question: 
Running the thread sanitizer seems like here the `snapshot_cv_` could be null (released in the destructor), should we wait for the job to be finished in the destructor? Or there is another way recommend to do it? @chenhao94 @xkszltl

Reply:
null? Either it's not init or it's deleted is an indication of data race. If so there should be a mutex around all its access.